docs/contributing/code_guide.rst

```diff
@@ -279,39 +279,138 @@ to edit the pull request by changing the corresponding branch on your
+
+  This refactoring strategy is appropriate for long functions where the
+  different steps can be cleanly separated from each other. This pattern
+  leads to functions that are shorter, more focused, more reusable, and
+  easier to test. The original function no longer includes
+  implementation details, and thus gives a high level view of what the
+  function is doing.
+
+  .. hint::
+
+     The `extract function refactoring pattern`_ might not be
+     appropriate if the different sections of code are intertwined with
+     each other (e.g., if both sections require the same intermediate
+     variables). An alternative in such cases would be to create a class
+     instead.
 
-* Separate subject from body with a blank line
+Error messages
+==============
 
-* Limit the subject line to 50 characters
+Error messages are a vital but underappreciated form of documentation.
+A good error message can help someone pinpoint the source of a problem
+in seconds, while a cryptic or missing error message can lead to hours
+of frustration.
 
-* Capitalize the subject line
+* Use error messages to indicate the source of the problem while
+  providing enough information for the user to fix it. When possible,
+  make it clear what the user should do next.
 
-* Do not end the subject line with a period
+* Include diagnostic information when appropriate.  If an occurred at
+  a single point in an array operation, including the coordinates of
+  where the error happened can help the user better understand the cause
+  of the error.
 
-* Use the imperative mood in the subject line
+* Write error messages that are concise when possible, as users often
+  skim or skip long error messages.
 
-* Wrap the body at 72 characters
+* Avoid including information that is irrelevant to the source of the
+  problem.
 
-* Use the body to explain what and why vs. how
+* Write error messages in language that is plain enough to be
+  understandable to someone who is undertaking their first research
+  project.
 
-Warnings and Exceptions
-=======================
+  - If necessary, technical information may be placed after a plain
+    language summary statement.
 
-* Debugging can be intensely frustrating when problems arise and the
-  associated error messages do not provide useful information on the
-  source of the problem.  Warnings and error messages must be helpful
-  enough for new users to quickly understand any problems that arise.
+  - Alternatively, an error message may reference a docstring or a page
+    in the narrative documentation.
 
-* "Errors should never pass silently."  Users should be notified when
-  problems arise by either issuing a warning or raising an exception.
+* Write error messages that are friendly, supportive, and helpful. Error
+  message should never be condescending or blame the user.
 
 * The exceptions raised by a method should be described in the
   method's docstring.  Documenting exceptions makes it easier for
```

Comment:
Do we have a doc guide rule on this?  This is one of the most difficult parts of writing a docstring.  And, I often find this part of the docstring is like a comment, always out of date.

Reply:
Hm...perhaps we could emulate the [numpydoc section](https://numpydoc.readthedocs.io/en/latest/format.html#raises) on this, and say that the `Raises` section should be used judiciously and only include non-obvious exceptions.  

- [ ] Do ↑
- [ ] Update the doc guide section on this too

> And, I often find this part of the docstring is like a comment, always out of date.

All too true...

Reply:
Updated this, and moved it to the doc guide as a means of procrastinating packing.